File: adventofcode/2020/day4/day4.rb
# frozen_string_literal: true

def read_input
  ARGF.read
end

FIELDS = {
  byr: lambda { |val|
    val = val.to_i
    val > 999 && val >= 1920 && val <= 2002
  },
  cid: ->(_) { true },
  ecl: ->(val) { %w[amb blu brn gry grn hzl oth].include?(val) },
  eyr: lambda { |val|
    val = val.to_i
    val > 999 && val >= 2020 && val <= 2030
  },
  hcl: ->(val) { val.match(/#([\da-f]{6})/) },
  hgt: lambda { |val|
    m = val.match(/^(\d+)(cm|in)/)
    return false unless m

    hgt = m[1].to_i
    dim = m[2]
    if dim == 'cm'
      hgt >= 150 && hgt <= 193
    else
      hgt >= 59 && hgt <= 76
    end
  },
  iyr: lambda { |val|
    val = val.to_i
    val > 999 && val >= 2010 && val <= 2020
  },
  pid: ->(val) { val.size == 9 }
}.freeze

EXCEPTIONS = [
  'cid'
].freeze

def parse_input(passports)
  passports.split("\n\n")
end

def parse_passport(passport)
  passport.split.map { |p| p.split(':').map(&:strip) }.to_h
end

def valid_passport?(passport)
  missing = FIELDS.keys.map(&:to_s) - passport.keys.sort
  missing.empty? || missing == EXCEPTIONS
end

def valid_passport2?(passport)
  valid_passport?(passport) && passport.all? do |k, v|
    FIELDS[k.to_sym].call(v)
  end
end

def part1(passports)
  parse_input(passports).select do |passport|
    valid_passport?(parse_passport(passport))
  end.size
end

def part2(passports)
  parse_input(passports).select do |passport|
    valid_passport2?(parse_passport(passport))
  end.size
end

passports = read_input
p part1(passports)
p part2(passports)

File: adventofcode/2020/day4/day4.hs
import Data.Char
import Data.List

partitionBy :: (a -> Bool) -> [a] -> [[a]]
partitionBy _ [] = []
partitionBy f (x:xs)
  | f x = partitionBy f xs
  | otherwise = (x:ts):partitionBy f rs
  where
    ts = takeWhile (not . f) xs
    rs = drop (length ts) xs

fields = ["byr", "ecl", "eyr", "hcl", "hgt", "iyr", "pid"]

isValid1 :: String -> Bool
isValid1 ps = null $ fields \\ kvp
  where pairs = partitionBy (== ' ') ps
        kv = map (partitionBy (== ':')) pairs
        kvp = map (\[k, _] -> k) kv

part1 :: [String] -> Int
part1 = length . filter isValid1

data PassField = Byr Int
               | Cid String
               | Ecl String
               | Eyr Int
               | Hcl String
               | Hgt Int String
               | Iyr Int
               | Pid String
               deriving (Show)

isValidField :: PassField -> Bool
isValidField (Byr x) = x `elem` [1920..2002]
isValidField (Cid x) = True
isValidField (Ecl x) = x `elem` ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
isValidField (Eyr x) = x `elem` [2020..2030]
isValidField (Hcl (x:xs))
  | x /= '#' = False
  | length xs /= 6 = False
  | otherwise = all (`elem` "abcdef0123456789") xs

isValidField (Hgt x "cm") = x `elem` [150..193]
isValidField (Hgt x "in") = x `elem` [59..76]
isValidField (Hgt x _) = False

isValidField (Iyr x) = x `elem` [2010..2020]
isValidField (Pid x)
  | length x /= 9 = False
  | otherwise = all isDigit x

isValid2 :: [PassField] -> Bool
isValid2 = all isValidField

parsePassField :: (String, String) -> PassField
parsePassField ("byr", x) = Byr (read x)
parsePassField ("ecl", x) = Ecl x
parsePassField ("eyr", x) = Eyr (read x)
parsePassField ("hcl", x) = Hcl x
parsePassField ("hgt", x) = Hgt (read val) dim
  where (val, dim) = span isDigit x
parsePassField ("iyr", x) = Iyr (read x)
parsePassField ("pid", x) = Pid x
parsePassField ("cid", x) = Cid x

parsePassport :: String -> [PassField]
parsePassport ps = map parsePassField kvp
  where pairs = partitionBy (== ' ') ps
        kv = map (partitionBy (== ':')) pairs 
        kvp = map (\[k, v] -> (k, v)) kv

part2 :: [String] -> Int
part2 = length . filter isValid2 . map parsePassport . filter isValid1

solve :: [String] -> String
solve ls = unlines $ map show [part1 ps, part2 ps]
  where ps = map unwords $ partitionBy (== "") ls

main :: IO ()
main = interact $ solve . lines
